adaptER-covid19 economics model

al626 opened the pull request:
Model to simulate economic effects of covid-19 in tandem with epidemiological model.

Still todo:
* test suite
* more api documentation

al626 commented on src/adapter_covid19/LICENSE:
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Hey Bryn,

This is slightly complicated, but this module can be Apache licensed as long as it doesn't use or interface with any of the GPL code that already exists in the repository (which is the case). The rest of the repository *can* use this module, as code licensed under Apache can be used by the GPL.

https://www.apache.org/licenses/GPL-compatibility.html

We talked with @sbalian and others about this, including getting a legal opinion, please let me know if this doesn't make sense or if it would be helpful to loop others in to explain.

Thanks

al626 replied:
Yep! It was just the first commit that we needed to be under Apache.

roberthinch commented:
Given that this is an add on to the epidemiology model and does not feed back in to it, then I am happy from our side. (final approval of the PR should come from Faculty who have been closer to this work)

sbalian commented on src/adapter_covid19/personal_insolvency.py:
```diff
@@ -0,0 +1,326 @@
+    corporate_bankruptcy: float
+
+
+@dataclass
```
Note model implemented differently from the others (using dataclass)

al626 replied:
So I've left this as it is, because there's no real reason for this to be a dataclass; this is something we've fixed in the next MR we're about to send across.

sbalian commented on src/setup.py:
```diff
@@ -27,19 +27,23 @@
 setup(
     name="covid19",
-    version="0.1",
+    version="0.2",
```
@brynmathias Happy with this? Have we settled with a convention?

brynmathias replied:
We're not settled on a convention yet, I will need to do so, for the moment this is fine.

sbalian commented on src/setup.py:
```diff
@@ -27,19 +27,23 @@
-        "COVID19/simulation",
+        'covid19',
     ],
     requires=[
```
Changing this to `install_requires` installed the dependencies, otherwise, they do not get installed with `pip` (as is in the makefile).

al626 commented on examples/example_run_simulation_with_economics.py:
```diff
@@ -0,0 +1,453 @@
+        output_dir=output_dir,
+    )
+
+    spread_processes = [
```
Done using `multiprocessing.Pool`.

Note that:

> processes is the number of worker processes to use. If processes is None then the number returned by os.cpu_count() is used.
(https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool)

al626 commented on src/adapter_covid19/constants.py:
```diff
@@ -0,0 +1,9 @@
+
+# Utilities
+UPPERCASE = set(string.ascii_uppercase)
+URL_REGEX = re.compile(r"^(https?|file)://")
```
removed

al626 commented on src/adapter_covid19/corporate_bankruptcy.py:
```diff
@@ -0,0 +1,220 @@
+class CorporateBankruptcyModel(NaiveCorporateBankruptcyModel):
+    def __init__(
+        self,
+        beta: float = 1 + np.random.rand(),
```
Ah, thanks for this, good spot!

al626 commented on src/adapter_covid19/corporate_bankruptcy.py:
```diff
@@ -0,0 +1,220 @@
+        :param consumption_cost_discount:
+        :return:
+        """
+        return {s: 1 for s in Sector}
```
So the idea behind this was that this class would be an easy way of running the economics simulator without running a corporate bankruptcy model (e.g. for testing purposes, or to see the what the model says without corporate bankruptcies), but decided to remove it

al626 commented on src/adapter_covid19/corporate_bankruptcy.py:
```diff
@@ -0,0 +1,220 @@
+        return {s: 1 for s in Sector}
+
+
+class CorporateBankruptcyModel(NaiveCorporateBankruptcyModel):
```
see below

al626 replied:
(removed)

al626 commented on src/adapter_covid19/corporate_bankruptcy.py:
```diff
@@ -0,0 +1,220 @@
+
+    def gdp_discount_factor(
+        self,
+        days_since_lockdown: int,
```
see below (removed)

al626 commented on src/adapter_covid19/corporate_bankruptcy.py:
```diff
@@ -0,0 +1,220 @@
+
+
+class NaiveCorporateBankruptcyModel:
+    def load(self, reader: Reader) -> None:
```
see below (removed)

al626 commented on src/adapter_covid19/corporate_bankruptcy.py:
```diff
@@ -0,0 +1,220 @@
+    def _get_mean_cash_buffer_days(
+        self,
+        lcap: bool,
+        gdp_discount: Optional[Mapping[Sector, float]],
```
> Optional[X] is equivalent to Union[X, None].

>Note that this is not the same concept as an optional argument, which is one that has a default. An optional argument with a default does not require the Optional qualifier on its type annotation just because it is optional

(https://docs.python.org/3/library/typing.html#typing.Optional)

al626 commented on src/adapter_covid19/enums.py:
```diff
@@ -0,0 +1,131 @@
+    }
+
+
+class Occupation(OrderedEnum):
```
removed

al626 commented on src/adapter_covid19/datasources.py:
```diff
@@ -0,0 +1,68 @@
+        return os.path.join(self.data_path, filename)
+
+    def load_csv(self, filename: str) -> pd.DataFrame:
+        return pd.read_csv(self._get_filepath(f"{filename}.csv"))
```
Appreciated this comment a lot. Much cleaner now

al626 commented on src/adapter_covid19/economics.py:
```diff
@@ -0,0 +1,112 @@
+        }
+
+
+class Economics:
```
Added some brief things, but we will be shortly submitting a PR with changed interfaces (and better documentation) so didn't flesh this out very much

al626 commented on src/adapter_covid19/model_docs.md:
```diff
@@ -0,0 +1,96 @@
+# AdaptER-Covid19
```
Moved to top level docs, but isn't this probably better mentioned from the `adaptER-covid19` readme?

brynmathias commented:
All good from my side, docker file change is minor

al626 commented:
Thanks Seto,

- [x] Rebased
- [x] Fixed runtime warning
- [x] Fixed path
- [ ] Will push some api docs in another MR - some interfaces are changing slightly so we will write docs for what our latest interfaces are.

sbalian commented:
Thanks @al626. @brynmathias looks good, go ahead and merge.
